db/version_builder.cc

```diff
@@ -514,10 +515,26 @@ class VersionBuilder::Rep {
     return Status::OK();
   }
 
-  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f) {
+  void MaybeInheritFile(VersionStorageInfo* vstorage, int level,
+                        FileMetaData* f) {
     if (levels_[level].deleted_files.count(f->fd.GetNumber()) > 0) {
       // f is to-be-deleted table file
       vstorage->RemoveCurrentStats(f);
+      deleted_base_files_[f->fd.GetNumber()] = f;
+    } else {
+      vstorage->AddFile(level, f, info_log_);
+    }
+  }
+
+  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f,
+                    uint64_t base_max_file_number) {
+    if (f->fd.GetNumber() <= base_max_file_number) {
+      auto it = deleted_base_files_.find(f->fd.GetNumber());
+      assert(it != deleted_base_files_.end());
+      vstorage->AddFile(level, it->second, info_log_);
```

Comment:
Do we need to bump ref count for the added file?

Reply:
`AddFile` will do that.